packages/vm/src/evm/evm.ts

```diff
@@ -129,120 +227,172 @@ export function VmErrorResult(error: VmError, gasUsed: bigint): ExecResult {
+    const evm = new this(opts)
+    await evm.init()
+    return evm
+  }
 
-    if (this._vm.DEBUG) {
-      debug('-'.repeat(100))
-      debug(`message checkpoint`)
-    }
+  constructor(opts: EVMOpts) {
+    super()
 
-    let result
-    if (this._vm.DEBUG) {
-      const { caller, gasLimit, to, value, delegatecall } = message
-      debug(
-        `New message caller=${caller} gasLimit=${gasLimit} to=${
-          to?.toString() ?? 'none'
-        } value=${value} delegatecall=${delegatecall ? 'yes' : 'no'}`
-      )
+    this._refund = BigInt(0)
+    this._transientStorage = new TransientStorage()
+
+    if (opts.common) {
+      this._common = opts.common
+    } else {
+      const DEFAULT_CHAIN = Chain.Mainnet
+      this._common = new Common({ chain: DEFAULT_CHAIN })
     }
-    if (message.to) {
-      if (this._vm.DEBUG) {
-        debug(`Message CALL execution (to: ${message.to})`)
+
+    // Supported EIPs
+    const supportedEIPs = [
+      1153, 1559, 2315, 2537, 2565, 2718, 2929, 2930, 3074, 3198, 3529, 3540, 3541, 3607, 3651,
+      3670, 3855, 3860, 4399,
+    ]
+
+    for (const eip of this._common.eips()) {
+      if (!supportedEIPs.includes(eip)) {
+        throw new Error(`EIP-${eip} is not supported by the VM`)
       }
-      result = await this._executeCall(message)
+    }
+
+    if (opts.vmState) {
+      this._state = opts.vmState
     } else {
-      if (this._vm.DEBUG) {
-        debug(`Message CREATE execution (to undefined)`)
+      const trie = new Trie()
+      const stateManager = new DefaultStateManager({
+        trie,
+        common: this._common,
+      })
+      this._state = new VmState({ common: this._common, stateManager })
+    }
+    this._blockchain = opts.blockchain ?? new (Blockchain as any)({ common: this._common })
+
+    this._allowUnlimitedContractSize = opts.allowUnlimitedContractSize ?? false
+    this._customOpcodes = opts.customOpcodes
+    this._customPrecompiles = opts.customPrecompiles
+
+    this._common.on('hardforkChanged', () => {
+      this.getActiveOpcodes()
+    })
+
+    // Initialize the opcode data
+    this.getActiveOpcodes()
+    this._precompiles = getActivePrecompiles(this._common, this._customPrecompiles)
+
+    if (this._common.isActivatedEIP(2537)) {
+      if (isBrowser()) {
+        throw new Error('EIP-2537 is currently not supported in browsers')
+      } else {
+        this._mcl = mcl
       }
-      result = await this._executeCreate(message)
     }
-    if (this._vm.DEBUG) {
-      const { gasUsed, exceptionError, returnValue } = result.execResult
-      debug(
-        `Received message execResult: [ gasUsed=${gasUsed} exceptionError=${
-          exceptionError ? `'${exceptionError.error}'` : 'none'
-        } returnValue=0x${short(returnValue)} gasRefund=${result.gasRefund ?? 0} ]`
-      )
+
+    // Safeguard if "process" is not available (browser)
+    if (process !== undefined && process.env.DEBUG) {
+      this.DEBUG = true
```

Comment:
seems like we have this `DEBUG` property in three places now, in the vm, in evm.ts, and interperer.ts, is there a way to simplify to just use one source of truth, setting and accessing it in one place? doing it additional times feels redundant

Reply:
I guess we DO need in VM and EVM if we separate the packages, but we should definitely unify in the EVM scope (so: evm.ts and interpreter.ts).

Reply:
Unified interpreter and evm to both use the evm debug prop.